File: src/qq_group_ops/services/export.py
"""信息导出服务。

导出属于高风险操作，必须做权限校验、脱敏和审计。
"""

from __future__ import annotations

import csv
from collections.abc import Iterable
from io import StringIO
from uuid import uuid4

from qq_group_ops.core.enums import AuditStatus
from qq_group_ops.core.models import AuditRecord
from qq_group_ops.services.activity import Activity, ActivityRegistration
from qq_group_ops.services.audit import AuditLog, InMemoryAuditLog
from qq_group_ops.services.permissions import PermissionService


def mask_identifier(value: str | None, *, keep: int = 1) -> str:
    """对 ID 做简单脱敏。"""

    if not value:
        return ""
    if len(value) <= keep * 2:
        return "*" * len(value)
    return f"{value[:keep]}***{value[-keep:]}"


class ExportService:
    """活动报名和审计日志导出。"""

    def __init__(
        self,
        permissions: PermissionService,
        audit_log: AuditLog | None = None,
    ) -> None:
        self._permissions = permissions
        self._audit_log = audit_log or InMemoryAuditLog()

    def export_activity_registrations_csv(
        self,
        *,
        actor_id: str,
        activity: Activity,
        registrations: Iterable[ActivityRegistration],
        mask_user_ids: bool = True,
    ) -> str:
        self._permissions.ensure(
            self._permissions.can_export_data(actor_id, activity.group_id),
            "permission denied: export activity registrations",
        )
        headers = [
            "registration_id",
            "activity_id",
            "group_id",
            "user_id",
            "display_name",
            "note",
            "created_at",
        ]
        rows: list[list[str]] = []
        count = 0
        for registration in registrations:
            count += 1
            rows.append(
                [
                    registration.registration_id,
                    registration.activity_id,
                    registration.group_id,
                    mask_identifier(registration.user_id) if mask_user_ids else registration.user_id,
                    registration.display_name,
                    registration.note,
                    registration.created_at.isoformat(),
                ]
            )
        content = self._render_csv(headers, rows)
        self._write_audit(
            actor_id=actor_id,
            group_id=activity.group_id,
            action="export_activity_registrations",
            count=count,
        )
        return content

    def export_audit_records_csv(
        self,
        *,
        actor_id: str,
        group_id: str,
        records: Iterable[AuditRecord],
        mask_user_ids: bool = True,
    ) -> str:
        self._permissions.ensure(
            self._permissions.can_export_data(actor_id, group_id),
            "permission denied: export audit records",
        )
        headers = [
            "record_id",
            "group_id",
            "actor_id",
            "target_user_id",
            "action",
            "status",
            "reason",
            "created_at",
        ]
        rows: list[list[str]] = []
        count = 0
        for record in records:
            count += 1
            rows.append(
                [
                    record.record_id,
                    record.group_id,
                    mask_identifier(record.actor_id) if mask_user_ids else record.actor_id,
                    mask_identifier(record.target_user_id) if mask_user_ids else (record.target_user_id or ""),
                    record.action,
                    record.status.value,
                    record.reason,
                    record.created_at.isoformat(),
                ]
            )
        content = self._render_csv(headers, rows)
        self._write_audit(
            actor_id=actor_id,
            group_id=group_id,
            action="export_audit_records",
            count=count,
        )
        return content

    @staticmethod
    def _render_csv(headers: list[str], rows: list[list[str]]) -> str:
        buffer = StringIO()
        writer = csv.writer(buffer)
        writer.writerow(headers)
        writer.writerows(rows)
        return buffer.getvalue()

    def _write_audit(self, *, actor_id: str, group_id: str, action: str, count: int) -> None:
        self._audit_log.append(
            AuditRecord(
                record_id=uuid4().hex,
                group_id=group_id,
                actor_id=actor_id,
                action=action,
                status=AuditStatus.EXECUTED,
                reason=f"rows={count}",
            )
        )

File: src/services/export.ts
import { randomUUID } from "node:crypto";

import { AuditStatus } from "../core/enums.js";
import { getLogger } from "../core/logger.js";
import type { AuditRecord } from "../core/models.js";
import { utcNow } from "../core/models.js";
import type { Activity, ActivityRegistration } from "./activity.js";
import type { AuditLog } from "./audit.js";
import { AuditLogStore } from "./audit.js";
import type { PermissionService } from "./permissions.js";

const log = getLogger("export");

export function maskIdentifier(value: string | undefined, keep = 1): string {
  if (!value) {
    return "";
  }
  if (value.length <= keep * 2) {
    return "*".repeat(value.length);
  }
  return `${value.slice(0, keep)}***${value.slice(-keep)}`;
}

export class ExportService {
  private readonly auditLog: AuditLog;

  public constructor(
    private readonly permissions: PermissionService,
    auditLog: AuditLog = new AuditLogStore(),
  ) {
    this.auditLog = auditLog;
  }

  public exportActivityRegistrationsCsv(
    actorId: string,
    activity: Activity,
    registrations: readonly ActivityRegistration[],
    maskUserIds = true,
  ): string {
    this.permissions.ensure(
      this.permissions.canExportData(actorId, activity.groupId),
      "permission denied: export activity registrations",
    );
    const headers = [
      "registration_id",
      "activity_id",
      "group_id",
      "user_id",
      "display_name",
      "note",
      "created_at",
    ];
    const rows = registrations.map((registration) => [
      registration.registrationId,
      registration.activityId,
      registration.groupId,
      maskUserIds ? maskIdentifier(registration.userId) : registration.userId,
      registration.displayName,
      registration.note,
      registration.createdAt.toISOString(),
    ]);
    const content = renderCsv(headers, rows);
    this.writeAudit(
      actorId,
      activity.groupId,
      "export_activity_registrations",
      registrations.length,
    );
    log.info("exported activity registrations", {
      actorId,
      groupId: activity.groupId,
      count: registrations.length,
    });
    return content;
  }

  public exportAuditRecordsCsv(
    actorId: string,
    groupId: string,
    records: readonly AuditRecord[],
    maskUserIds = true,
  ): string {
    this.permissions.ensure(
      this.permissions.canExportData(actorId, groupId),
      "permission denied: export audit records",
    );
    const headers = [
      "record_id",
      "group_id",
      "actor_id",
      "target_user_id",
      "action",
      "status",
      "reason",
      "created_at",
    ];
    const rows = records.map((record) => [
      record.recordId,
      record.groupId,
      maskUserIds ? maskIdentifier(record.actorId) : record.actorId,
      maskUserIds ? maskIdentifier(record.targetUserId) : (record.targetUserId ?? ""),
      record.action,
      record.status,
      record.reason,
      record.createdAt.toISOString(),
    ]);
    const content = renderCsv(headers, rows);
    this.writeAudit(actorId, groupId, "export_audit_records", records.length);
    log.info("exported audit records", { actorId, groupId, count: records.length });
    return content;
  }

  private writeAudit(
    actorId: string,
    groupId: string,
    action: string,
    count: number,
  ): void {
    this.auditLog.append({
      recordId: randomUUID(),
      groupId,
      actorId,
      action,
      status: AuditStatus.Executed,
      reason: `rows=${count}`,
      createdAt: utcNow(),
    });
  }
}

function renderCsv(headers: readonly string[], rows: readonly (readonly string[])[]): string {
  return [headers, ...rows]
    .map((row) => row.map(csvEscape).join(","))
    .join("\n")
    .concat("\n");
}

function csvEscape(value: string): string {
  if (/[",\n\r]/.test(value)) {
    return `"${value.replaceAll('"', '""')}"`;
  }
  return value;
}
